User-requested usability features for batch processing

tbowers7 opened the pull request:
There are two usability changes included here.

1. Allow the user to turn off writing Spec2d files for science frames.

While this should not be used for the majority of cases, when a user is reducing thousands of time-resolved spectra, it may be useful to turn off the writing of 2D spectra to save disk space.  A warning is added to the new parameter indicating that this should only be used once the user is satisfied that the reduction is working properly and that the 2D spectrum is no longer needed for debugging and quality assurance purposes.

Invoke this option by adding the following to the parameter block of the reduction file:
```
[scienceframe]
    [[process]]
        skip_write_2d = True
```


2. Allow the the user to pass in a file containing additional parameters to ``pypeit_setup`` that will be included in the resulting PypeIt Reduction File.

For projects containing multiple nights' worth of data that use a small number of spectrograph-divergent parameters (e.g., object-finding or flexure parameters), it is helpful if ``pypeit_setup`` could automatically include these in the reduction file.  A new command-line argument to ``pypeit_setup`` is added (``"-p", "--param_block_file"``) that takes the name of a file containing the pre-formatted parameters to be added.

For example, the file "xtra_params.txt" might contain the following lines:
```
[calibrations]
  [[slitedges]]
    minimum_slit_length = 45
  [[flatfield]]
    pixelflat_min_wave = 4200
[reduce]
  [[findobj]]
    find_fwhm = 6.0
  [[extraction]]
    use_user_fwhm = True
```
  Use of these lines would be done with, _e.g._,:
```
  $ pypeit_setup -s ldt_deveny -c A -p xtra_params.txt
```

---

Other minor updates include ``os`` --> ``pathlib`` and typing/docstring updates in ``pypeit.py`` somewhat connected to the above changes.  These are code style changes and should not affect functionality at all.

---

Cloud Dev-Suite run in progress... will post results when complete.

tbowers7 committed:
Usability changes requested by user

There are two usability changes included here.

1. Allow the user to turn off writing Spec2d files for sciece frames.

  While this should not be used for the majority of cases, when a
  user is reducing thousands of time-resolved spectra, it may be
  useful to turn off the writing of 2D spectra to save on disk
  space.  A warning is added to the new parameter indicating that
  this should only be used once the user is satisified that the
  reduction is working properly and that the 2D spectrum is no
  longer needed for debugging and quality assurance purposes.

  Invoke this option by adding the following to the parameter block
  of the reduction file:

[scienceframe]
    [[process]]
        skip_write_2d = True

2. All the the user to pass in a file containing additional parameters
   to ``pypeit_setup`` that will be included in the resulting PypeIt
   Reduction File.

  For projects containing multiple nights' worth of data that use a
  small number of spectrograph-divergent parameters (e.g., object-
  finding or flexure parameters), it is helpful if ``pypeit_setup``
  could automatically include these in the reduction file.  A new
  command-line argument to ``pypeit_setup`` is added ("-p",
  "--param_block_file") that takes the name of a file containing
  the pre-formatted parameters to be added.

  For example, the file "xtra_params.txt" might contain the
  following lines:

[calibrations]
  [[slitedges]]
    minimum_slit_length = 45
  [[flatfield]]
    pixelflat_min_wave = 4200
[reduce]
  [[findobj]]
    find_fwhm = 6.0
  [[extraction]]
    use_user_fwhm = True

  Use of these lines would be done with, e.g.,:

  $ pypeit_setup -s ldt_deveny -c A -p xtra_params.txt

tbowers7 committed:
Merge branch 'develop' into user_request

tbowers7 committed:
Merge branch 'develop' into user_request

tbowers7 committed:
Merge branch 'develop' into user_request

tbowers7 committed:
Mostly cleaning

Adding some docstrings, moving os.path -> pathlib where appropriate,
remove unused imports.  Also, warning message if the --param_block_file
optional input to pypeit_setup does not point to an actual file.

	modified:   pypeit/datamodel.py
	modified:   pypeit/pypeit.py
	modified:   pypeit/pypeitsetup.py
	modified:   pypeit/scripts/setup.py

tbowers7 committed:
More docstring, adjust new method name

	modified:   pypeit/metadata.py
	modified:   pypeit/pypeit.py
	modified:   pypeit/pypeitsetup.py
	modified:   pypeit/scripts/setup.py

tbowers7 committed:
updated docs

tbowers7 committed:
Release Notes

	modified:   doc/releases/1.17.5dev.rst

tbowers7 commented:
This cloud run was done against the `develop` branch of the Dev-Suite.  I'm going to need some help interpreting what all is going on here.  I think the LBT/MODS failures are expected and will be fixed soon, but the rest?  I don't think the changes in this PR should be related to `sensfunc` failures.

```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS FAILED  1 failed, 258 passed, 445 warnings in 855.11s (0:14:15) ---
--- PYTEST UNIT TESTS FAILED  3 failed, 149 passed, 1584 warnings in 1411.02s (0:23:31) ---
--- PYTEST VET TESTS FAILED  6 failed, 61 passed, 1160 warnings in 9451.03s (2:37:31) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 15/251 TESTS  ---
Failed tests:
    lbt_mods/MODS2R_Longslit pypeit
    gemini_gmos/GS_HAM_R400_860 pypeit_sensfunc
    lbt_mods/MODS1R_Longslit pypeit
    keck_mosfire/Y_long pypeit_sensfunc
    keck_deimos/900ZD_LVM_5500 pypeit_sensfunc
    gemini_gmos/GS_HAM_R400_700 pypeit_sensfunc
    keck_lris_red_mark4/long_600_10000_d680 pypeit_sensfunc
    shane_kast_red/600_7500_d55_ret pypeit_sensfunc
    vlt_xshooter/VIS_1x1_LTT3218 pypeit_sensfunc
    gemini_gnirs_echelle/32_SB_SXD pypeit_sensfunc
    vlt_xshooter/VIS_1x1_Feige110 pypeit_sensfunc
    vlt_xshooter/NIR_Feige110 pypeit_sensfunc
    vlt_xshooter/UVB_1x1_Feige110 pypeit_sensfunc
    vlt_xshooter/NIR_LTT3218 pypeit_sensfunc
    keck_nires/ABpat_wstandard pypeit_ql
Skipped tests:
    gemini_gmos/GS_HAM_R400_860 pypeit_flux_setup
    gemini_gmos/GS_HAM_R400_860 pypeit_flux
    keck_mosfire/Y_long pypeit_ql
    keck_deimos/900ZD_LVM_5500 pypeit_flux
    gemini_gmos/GS_HAM_R400_700 pypeit_flux
    gemini_gmos/GS_HAM_R400_700 pypeit_coadd_1dspec
    gemini_gmos/GS_HAM_R400_700 pypeit_tellfit
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
    shane_kast_red/600_7500_d55_ret pypeit_flux
    gemini_gnirs_echelle/32_SB_SXD pypeit_flux_setup
    gemini_gnirs_echelle/32_SB_SXD pypeit_flux
    gemini_gnirs_echelle/32_SB_SXD pypeit_coadd_1dspec
    gemini_gnirs_echelle/32_SB_SXD pypeit_coadd_2dspec
    gemini_gnirs_echelle/32_SB_SXD pypeit_tellfit
    keck_nires/ABpat_wstandard pypeit_ql
    keck_nires/ABpat_wstandard pypeit_ql
Testing Started at 2025-07-10T17:25:27.543407
Testing Completed at 2025-07-11T14:20:51.944044
Total Time: 20:55:24.400637
```

tbowers7 commented:
UPDATE: 

The PypeIt and Dev-Suite unit test failures were caused by a failure to download a telluric grid file from the cloud... and it looks like the various `sensfunc` tests failed with the same timeout error looking for the required tellgrid file.

kbwestfall commented:
Yeah, these timeouts are annoying for testing.  I would first confirm that all the unit tests pass locally (you can just run `pytest .` within each relevant directory).  Then, we might as well run the tests again.  Yes, the LBT/MODS tests will continue to fail until we merge #1927 .

tbowers7 commented:
> Yeah, these timeouts are annoying for testing. 

Would it make sense to -- for the purposes of cloud testing -- allow the test file to specify a longer timeout?

Right now, the timeout is hardwired at 10s ([pypeit/cache.py:L252](https://github.com/tbowers7/PypeIt/blob/cf57913e1ac7354c9422e3cd1984fbd13a62ef70/pypeit/cache.py#L252)), but extending this in the test might allow for fewer "failed runs".

Alternatively, could we add the TellPCA files to the `gen_kube_devsuite` script ([nautilus/gen_kube_devsuite:L100](https://github.com/pypeit/PypeIt-development-suite/blob/726a14c03576652b095816dac7d13df721bf1921/nautilus/gen_kube_devsuite#L100-L106)) so that they are copied over at the start of the Dev-Suite run also?

tbowers7 committed:
Fix a boo boo

Missed a consequence of my ``os`` --> ``pathlib`` changes.

	modified:   pypeit/scripts/ql.py

tbowers7 committed:
Temp fix for cloud devsuite run

Extend the timeout to allow for remote services to respond when
overloaded.

	modified:   pypeit/cache.py

tbowers7 commented:
Dev Suite passes now except for the LBT/MODS failures.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  259 passed, 445 warnings in 330.79s (0:05:30) ---
--- PYTEST UNIT TESTS PASSED  152 passed, 136817 warnings in 800.79s (0:13:20) ---
--- PYTEST VET TESTS FAILED  1 failed, 66 passed, 1194 warnings in 4639.03s (1:17:19) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/267 TESTS  ---
Failed tests:
    lbt_mods/MODS2R_Longslit pypeit
    lbt_mods/MODS1R_Longslit pypeit
Skipped tests:
Testing Started at 2025-07-11T22:27:13.911807
Testing Completed at 2025-07-12T08:54:14.467954
Total Time: 10:27:00.556147
```

To help deal with the timeout errors, had `nautilus` container also copy over the TellPCA files (https://github.com/pypeit/PypeIt-development-suite/pull/362) -- so ran against that branch (rather than `develop`).  I also did a little cleaning of pytest warnings, and one change exposed a VET test failure that wasn't being caught (https://github.com/pypeit/PypeIt/issues/1951).

tbowers7 committed:
Revert "Temp fix for cloud devsuite run"

	modified:   pypeit/cache.py

tbowers7 committed:
Merge branch 'develop' into user_request

badpandabear commented on doc/help/pypeit_setup.rst:
```diff
@@ -74,5 +74,9 @@
                             tell pypeit_obslog to remove them. (default: False)
+      -p, --param_block_file PARAM_BLOCK_FILE
+                            File containing the additional PypeIt user parameters to
+                            be added to the parameter block of the geerated
```
Should be generated

badpandabear commented on pypeit/scripts/setup.py:
```diff
@@ -58,6 +58,9 @@ def get_parser(cls, width=None):
                                  'them.')
+        parser.add_argument('-p', '--param_block_file', default=None, type=str,
+                            help='File containing the additional PypeIt user parameters to be '
+                                 'added to the parameter block of the geerated reduction file')
```
Should be generated

kbwestfall commented on pypeit/datamodel.py:
```diff
@@ -1089,7 +1089,7 @@ def _parse(cls, hdu, ext=None, ext_pseudo=None, transpose_table_arrays=False,
         #   - Hack to force native byte ordering
         for key in _d:
-            if isinstance(_d[key], np.chararray):
+            if isinstance(_d[key], np.char.chararray):
```
Thanks for this fix!  I was getting deprecation warnings about this, but wasn't sure about the correct replacement.

kbwestfall commented on pypeit/pypeit.py:
```diff
@@ -191,37 +190,51 @@ def build_qa(self):
+                Is this a 2D science frame?  (Default: False)
+
+        Returns:
+            `Path`_: The spectrum filename
```
Thanks for the documentation updates!

kbwestfall reviewed (approved):
Looks good.  Thanks for the doc updates.

I realize that I'm likely in the minority on this, but I continue to think we should avoid including default values in doc strings.  The docs show the calling sequence with the true defaults, and explicitly adding the defaults in the docstring means we have to keep track of it when the default changes.  I'd rather people have to scroll to the calling sequence to see the default value than have the defaults be different in different places.  Rant ended...

I'm approving, but can you add a test for `pypeit_setup` that uses the new configuration code?  This may need to be a unit test in the dev-suite.

tbowers7 commented:
> I realize that I'm likely in the minority on this, but I continue to think we should avoid including default values in doc strings. The docs show the calling sequence with the true defaults, and explicitly adding the defaults in the docstring means we have to keep track of it when the default changes. I'd rather people have to scroll to the calling sequence to see the default value than have the defaults be different in different places. Rant ended...

I have no particular idealogical stance on this.  The template I use in VSCode includes a statement of default values...

I think I agree with your stance on the topic.

tbowers7 commented:
> I'm approving, but can you add a test for `pypeit_setup` that uses the new configuration code? This may need to be a unit test in the dev-suite.

Yes.  Didn't see an obvious place for a unit test in the main PypeIt repo... will investigate the dev-suite unit tests.

kbwestfall commented:
> > I'm approving, but can you add a test for `pypeit_setup` that uses the new configuration code? This may need to be a unit test in the dev-suite.
> 
> Yes. Didn't see an obvious place for a unit test in the main PypeIt repo... will investigate the dev-suite unit tests.

I think `unit_tests/test_setups.py` is likely where this should go.

tbowers7 committed:
PR Comments

Fixed spelling error, and removed docstring "Default: " values.  Fortunately,
there were not many of the latter in the whole repo, so now cleansed.

	modified:   doc/help/pypeit_setup.rst
	modified:   doc/releases/1.17.5dev.rst
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/core/wavecal/spectrographs/templ_ldt_deveny.py
	modified:   pypeit/core/wavecal/spectrographs/templ_mdm_modspec_echelle.py
	modified:   pypeit/core/wavecal/spectrographs/templ_mmt_binospec.py
	modified:   pypeit/core/wavecal/spectrographs/templ_mmt_bluechannel.py
	modified:   pypeit/pypeit.py
	modified:   pypeit/scripts/setup.py

tbowers7 committed:
Add "important" directive in docstring

	modified:   pypeit/pypeitsetup.py

tbowers7 commented:
Unit test for the new `-p` option to `pypeit_setup` has been added in the dev-suite unit tests with https://github.com/pypeit/PypeIt-development-suite/pull/362.

tbowers7 committed:
Merge branch 'develop' into user_request

tbowers7 commented:
New Dev-Suite unit tests pass (154 tests now compared to the 152 previously).  The one PypeIt unit test failure was a timeout error downloading a TellGrid file.

```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS FAILED  1 failed, 258 passed, 446 warnings in 901.83s (0:15:01) ---
--- PYTEST UNIT TESTS PASSED  154 passed, 136818 warnings in 1560.52s (0:26:00) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 0/0 TESTS  ---
Testing Started at 2025-07-17T12:53:10.414768
Testing Completed at 2025-07-17T13:34:19.305569
Total Time: 0:41:08.890801
```
